Sync Vue Nodes Right Click menu with full ContextMenu

Myestery opened the pull request:
This pull request introduces several improvements to the node context menu in the graph selection toolbox, focusing on enhanced usability, accessibility, and maintainability. The most significant changes are the addition of a search/filter feature for menu options, improved handling and display of disabled and categorized menu items, and better viewport-aware positioning for popovers and submenus.

**Menu Usability & Filtering:**
* Added a search input at the top of the node options menu, allowing users to filter menu items by label. Filtering contextually hides dividers and category labels, and resets/focuses the search field when the menu is opened or closed. (`NodeOptions.vue`, `useMoreOptionsMenu.ts`) [[1]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eL14-R43) [[2]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eL75-R141) [[3]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eR334-R337) [[4]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eR354-R361) [[5]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eR373-R374)

**Menu Item Display & Accessibility:**
* Improved menu item rendering to support category labels and visually distinguish disabled options. Disabled options are now styled and prevented from triggering actions when clicked. (`MenuOptionItem.vue`, `SubmenuPopover.vue`, `useMoreOptionsMenu.ts`) [[1]](diffhunk://#diff-c5e6830e63e2441d2dc70d2ecf7c9b56d0a93821f827e9c5377fc10ae6016f18R6-R20) [[2]](diffhunk://#diff-c5e6830e63e2441d2dc70d2ecf7c9b56d0a93821f827e9c5377fc10ae6016f18R71-R73) [[3]](diffhunk://#diff-5ef501a9d46fd815df4f5a5cffac32e777291134897af7c9ee56bb011099e57aR22-R30) [[4]](diffhunk://#diff-5ef501a9d46fd815df4f5a5cffac32e777291134897af7c9ee56bb011099e57aL72-R122) [[5]](diffhunk://#diff-ae87bdb1e06725eb19b8d0fc82ec40a5f8ca831a6632767cc5d214fc903b89b6L18-R36)
* Extended the menu option type definitions to include `category` and `disabled` flags, and added a `source` field for distinguishing between hardcoded and dynamic options. (`useMoreOptionsMenu.ts`) [[1]](diffhunk://#diff-ae87bdb1e06725eb19b8d0fc82ec40a5f8ca831a6632767cc5d214fc903b89b6L18-R36) [[2]](diffhunk://#diff-ae87bdb1e06725eb19b8d0fc82ec40a5f8ca831a6632767cc5d214fc903b89b6R85-R97)

**Popover & Submenu Positioning:**
* Refactored popover positioning logic to use a viewport-aware utility, preventing menus and submenus from overflowing the visible area. Submenus now dock to the bottom of the viewport if they would otherwise overflow. (`NodeOptions.vue`, `SubmenuPopover.vue`) [[1]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eR183-R207) [[2]](diffhunk://#diff-5ef501a9d46fd815df4f5a5cffac32e777291134897af7c9ee56bb011099e57aL72-R122)

**Codebase Maintenance:**
* Updated imports and removed unused code related to canvas interactions, and improved composable structure for menu options. (`NodeOptions.vue`, `useMoreOptionsMenu.ts`) [[1]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eR60) [[2]](diffhunk://#diff-e0dbd5e37efd2c79e7317415455340b0dd150b758077b170a663f67d2453605eL56-R89) [[3]](diffhunk://#diff-ae87bdb1e06725eb19b8d0fc82ec40a5f8ca831a6632767cc5d214fc903b89b6R5-R11) [[4]](diffhunk://#diff-ae87bdb1e06725eb19b8d0fc82ec40a5f8ca831a6632767cc5d214fc903b89b6R115-R133)

These changes collectively enhance the user experience for interacting with node options, making the menu more navigable, accessible, and robust.


https://github.com/user-attachments/assets/cb3c9aca-45ad-4a87-b571-26d6d21a0559

Myestery committed:
right click vue nodes with contextmenu WIP

Myestery committed:
[feat] Add disabled state support for context menu items

- Add disabled property to MenuOption and SubMenuOption interfaces
- Apply semantic token styling for disabled items (text-node-icon-disabled)
- Prevent interactions on disabled menu and submenu items
- Use cursor-not-allowed and pointer-events-none for disabled state
- Pass through disabled flag from LiteGraph menu items to Vue components

Myestery committed:
feat: add Extensions translation key to context menu

Add translation key for the Extensions section label in the context menu.
This allows the Extensions category header to be properly localized.

Myestery committed:
feat: add category type support to menu items

Add support for non-clickable category labels in the context menu.
Category items are displayed as uppercase labels with secondary text color
and proper translation support via contextMenu namespace.

This enables grouping menu items under section headers like "Extensions".

Myestery committed:
feat: implement menu ordering and extension categorization system

Add comprehensive menu ordering system to ensure consistent context menu layout:
- Define MENU_ORDER constant with 5 logical sections
- Implement automatic section-based divider insertion
- Separate core items from extension items
- Add blacklist for unwanted duplicate items (Colors, Shapes, Title, Mode, etc.)
- Preserve Vue hardcoded options over LiteGraph items during deduplication
- Add source tracking (litegraph vs vue) for precedence handling

Menu structure:
1. Basic operations (Rename, Copy, Duplicate)
2. Node actions (Run Branch, Pin, Bypass, Mute)
3. Structure operations (Convert to Subgraph, Frame selection, Minimize Node)
4. Node properties (Node Info, Color)
5. Node-specific operations (Image operations)
6. Extensions section (non-core items)
7. Delete (always at bottom)

The system ensures proper ordering regardless of whether items come from
LiteGraph or Vue hardcoded menus, with Extensions clearly separated.

Myestery committed:
feat: restructure context menu sections with proper ordering

Reorder menu sections to follow the specification:
1. Basic operations: Rename, Copy, Duplicate
2. Node actions: Run Branch, Pin, Bypass, Mute (for groups)
3. Structure operations: Convert to Subgraph, Frame selection, Minimize Node
4. Node properties: Node Info, Color (Shape removed)
5. Node-specific operations: Image operations only

Changes:
- Add explicit dividers after each section
- Remove Shape, Alignment, and Adjust Size options
- Merge LiteGraph and Vue options for single node selection
- Add source tracking (vue) for precedence during deduplication
- Add comprehensive logging for debugging menu construction

For single node selection, both LiteGraph and Vue options are merged
with Vue options taking precedence, then structured via buildStructuredMenu.

Myestery committed:
feat: add search functionality to context menu

Add real-time search filtering to the node context menu:
- Search input with icon at the top of the menu
- Case-insensitive filtering by menu item label
- Auto-focus on menu open
- Clear search on Escape key or menu close
- Smart divider handling (removes unnecessary dividers during search)
- Hides Extensions category label when searching

The search provides instant feedback as users type, making it easier
to find specific menu items in long context menus.

Myestery committed:
fix: move Resize and Clone to structure operations section

Move Resize and Clone menu items from fallback section to Section 3
(Structure operations) so they appear alongside Convert to Subgraph,
Frame selection, and Minimize Node.

Updated section boundaries:
- Section 3 now spans indices 9-15 (was 9-13)
- Section 4 now spans indices 16-17 (was 14-15)

This ensures Resize and Clone appear under the same divider group
as other structure-related operations.

Myestery committed:
WIP scroll

Myestery committed:
feat: add viewport-aware positioning to prevent menu overflow

When right-clicking near the bottom of the viewport, the context menu now docks to the bottom edge instead of floating off-screen, ensuring it remains accessible and scrollable.

Myestery committed:
refactor: make position types internal to prevent unused export warnings

Myestery committed:
refactor: remove unused menuOptionExists function

DrJKL commented on src/components/graph/selectionToolbox/NodeOptions.vue:
```diff
@@ -53,13 +74,19 @@ import type {
-import { useCanvasInteractions } from '@/renderer/core/canvas/useCanvasInteractions'
+import { calculateMenuPosition } from '@/composables/graph/useViewportAwareMenuPositioning'
+
+// import { useCanvasInteractions } from '@/renderer/core/canvas/useCanvasInteractions'
```
```suggestion
```

Myestery committed:
feat: add fuzzy search with debouncing to context menu

Replaces simple substring matching with Fuse.js fuzzy search for better
user experience. Adds 300ms debouncing to reduce processing overhead.
Includes debug logging for troubleshooting.

Myestery committed:
refactor: remove debug console logs from NodeOptions

Myestery committed:
Merge remote-tracking branch 'origin/main' into right-click-vue-node-contextmenu

Myestery committed:
fix searchbox color and add empty state

Myestery committed:
fix: adjust badge height and update search placeholder for better UX

Myestery committed:
fix: update search input autofocus behavior and add mobile viewport handling

Myestery committed:
refactor: remove debug logging from context menu functions for cleaner code

Myestery committed:
Merge branch 'main' into right-click-vue-node-contextmenu

Myestery committed:
[feat] add shape option to node context menu and improve CSS positioning

Myestery committed:
refactor: Update submenu repositioning logic to use a direct content ref.

DrJKL commented on src/composables/graph/useViewportAwareMenuPositioning.ts:
How much of this is necessary because we're not using a [good existing component](https://reka-ui.com/docs/components/context-menu#features)?

DrJKL commented on src/components/graph/selectionToolbox/NodeOptions.vue:
```diff
@@ -22,15 +22,43 @@
+              :placeholder="t('contextMenu.Search')"
+              class="w-full rounded-lg border-0 focus:border py-2 pl-9 pr-3 text-sm text-text-primary placeholder-text-secondary focus:outline-none bg-secondary-background"
+              @keydown.escape="clearSearch"
+            />
```
Do we not have a component for this?

Myestery committed:
fix: dock menu to top of viewport when trigger is above

Myestery committed:
style: Replace `text-[12px]` with `text-xs` utility

Myestery committed:
feat: allow null values in context menu options and update related type definitions and method signatures.

Myestery committed:
feat: Replace custom search input with  component in NodeOptions.

Myestery committed:
Merge branch 'main' into right-click-vue-node-contextmenu

Myestery committed:
[refactor] Replace NodeOptions with PrimeVue ContextMenu

- Create NodeContextMenu.vue using pure PrimeVue ContextMenu
- Remove custom positioning logic, let PrimeVue handle it
- Delete NodeOptions.vue and MenuOptionItem.vue
- Mount NodeContextMenu in SelectionToolbox
- Update test stubs

Myestery committed:
[refactor] Rename SubmenuPopover to ColorPickerMenu

- Rename SubmenuPopover.vue to ColorPickerMenu.vue
- Add isColorPicker prop to MenuOption interface
- Use isColorPicker flag instead of checking submenu colors
- Fix onClickOutside immediate close with ignore option

Myestery committed:
[chore] Remove unused positioning files and exports

- Delete useSubmenuPositioning.ts (no longer used)
- Delete useViewportAwareMenuPositioning.ts (no longer used)
- Remove unused exports from useSelectionToolboxPosition.ts

Myestery committed:
refactor: replace specific hover background colors with theme variable in ColorPickerMenu

Myestery committed:
style: enhance shortcut display in the node context menu.

Myestery committed:
feat: ensure context menu's `isOpen` state is correctly updated on show and adjust submenu test interaction to hover

Myestery committed:
test: add 500ms page timeout

Myestery committed:
fix: Correct 'Open in Mask Editor' text and remove redundant `isOpen` initialization in node context menu.

Myestery commented on src/components/graph/selectionToolbox/NodeOptions.vue:
it is now accesible

Myestery committed:
feat: Constrain node context menu dimensions with overflow scrolling and improve legacy item extraction robustness.

Myestery committed:
Merge branch 'main' into right-click-vue-node-contextmenu

DrJKL commented:
@coderabbitai full review
